x509/cert.go

```diff
@@ -30,3 +31,30 @@ func ParseCertificatePEM(data []byte) ([]*x509.Certificate, error) {
 	}
 	return certs, nil
 }
+
+// Validates a certificate chain from leaf to root. Expects certChain input to be ordered
+// Returns the full ordered chain: [leaf, intermediates..., root]
+func buildAndValidateCertChain(certChain []*x509.Certificate) error {
+	if len(certChain) < 2 {
+		return errors.New("certificate chain must contain at least two certificates")
+	}
+
+	for i, cert := range certChain {
+		if i == len(certChain)-1 {
+			if !isRootCertificate(cert) {
+				return errors.New("certificate chain does not end in a root")
+			}
+		} else {
+			if err := cert.CheckSignatureFrom(certChain[i+1]); err != nil || isRootCertificate(cert) {
+				return errors.New("certificate chain does not validate sequentially from signing cert to root")
```

Comment:
This error message can be bit more descriptive about what failed in the validation. 

```
signature on certificate "<subject>" is not issued by "<subject>"
```